### DISCUSSION TITLE
org-ql-search: sort headers according to timestamp globally

### DISCUSSION THREAD

**Discussion by timmli** [alphapapa/org-ql#416] [Q&A] [answered]
I have several `org-agenda-files` and I'm trying to sort the contained appointments/events depending on their timestamp/date (from newest to oldest) with this code:
```
(org-ql-search
  (org-agenda-files)
  '(tags "event" "appointment")
  :sort '(date))
```
Unfortunately, appointments below the same header seem to be grouped together, thus the sorting is not global.
How can I achieve a global sorting of these appointments, regardless of their position in an Org file?

**Comment by alphapapa** [comment]
`org-ql-search` returns a list of matching entries (i.e. headings), not a list of appointments or events or timestamps.  If an entry contains multiple timestamps...

**Reply by timmli** [reply]
An "appointment" is, in my case, a header with tag `:appointment:` and usually an active timestamp. If it contains more than one timestamp, it probably should appear more than once in the list ....?

Here is an example for an Org file:
```
* Top Header 1
** <2024-02-28 Wed> Appointment 1.1                            :appointment:
** <2024-02-29 Thu> Appointment 1.2                            :appointment:

* Top Header 2
** <2024-02-29 Thu> Appointment 2.2                            :appointment:
** <2024-02-28 Wed> Appointment 2.1                            :appointment:
```
Using the code above generates the following list, which just mirrors the structure of the file:
```
  <2024-02-28 Wed> Appointment 1.1                                                                                  :appointment:
  <2024-02-29 Thu> Appointment 1.2                                                                                  :appointment:
  <2024-02-29 Thu> Appointment 2.2                                                                                  :appointment:
  <2024-02-28 Wed> Appointment 2.1                                                                                  :appointment:
 (0:00)
```
Meanwhile, I found out that `:sort '(date)` actually only takes the DEADLINE or SCHEDULED property into account. So my code obviously cannot work as intented. :1st_place_medal: 

Instead of using org-ql, I now customize the agenda view in the following way, which pretty much does the desired job:
```
(setq org-agenda-custom-commands
      `(("e" "Events with timestamp"
         ((tags "event+TIMESTAMP>=\"<now>\""
                ((org-agenda-overriding-header "Future events")
                 (org-agenda-sorting-strategy '((tags ts-down)))
                 (org-agenda-prefix-format '((tags . " %i %-12:c %?-12t")))
                 ))
          (tags "event+TIMESTAMP<=\"<now>\""
                ((org-agenda-overriding-header "Past events")
                 (org-agenda-sorting-strategy '((tags ts-down)))
                 (org-agenda-prefix-format '((tags . " %i %-12:c %?-12t")))
                 ))))))
```
Still it would be interesting to see how this could be achieved with org-ql.

**Comment by alphapapa** [comment, marked as answer]
> Still it would be interesting to see how this could be achieved with org-ql.

You would need to provide a sort function that collects timestamps in an entry and chooses one (e.g. the earliest or latest) to compare with the one chosen from the other entry.  You can see some similar code here: https://github.com/alphapapa/org-ql/blob/e41fe9018a4699532ec875bedddc9746f8e362aa/org-ql.el#L2325-L2388

It's intended to provide more sorting functions like that in the future.  It's just a matter of finding the time to work on these ideas.

See also some related code here: https://github.com/alphapapa/unpackaged.el#refile-to-datetree-file-using-earliestlatest-timestamp-in-entry
